File: 3200-maximum-height-of-a-triangle/3200-maximum-height-of-a-triangle.java
class Solution {
    public int maxHeightOfTriangle(int red, int blue) {
        return Math.max(helper(red, blue, 1, true), helper(red, blue, 1, false));
    }
    
    private int helper(int red, int blue, int current, boolean isRed){
        if(isRed){
            if(red < current) return 0;
            return 1 + helper(red - current, blue, current+1, !isRed);
        }else{
            if(blue < current) return 0;
            return 1 + helper(red, blue - current, current+1, !isRed);
        }
    }
}

File: 3200-maximum-height-of-a-triangle/3200-maximum-height-of-a-triangle.kt
class Solution {
    fun maxHeightOfTriangle(red: Int, blue: Int): Int {
        return Math.max(helper(red, blue, false, 1), helper(red, blue, true, 1))
    }
    
    private fun helper(red: Int, blue: Int, isBlue: Boolean, level: Int): Int{
        if(isBlue){
            if(blue < level) return 0
            return 1 + helper(red, blue-level, false, level+1)
        }else{
            if(red < level) return 0
            return 1 + helper(red-level, blue, true, level+1)
        }
    }
}